h2o-extensions/mojo-pipeline/src/main/java/hex/mojopipeline/H2OAssemblyToMojoPipelineConverter.java

```diff
@@ -0,0 +1,285 @@
+            if (!stage.isInPlace() &&stage.getNewNames().length > 0) {
+                Column[] tempColumns = convertColumns(stage.getNewNames(), stage.getNewTypes());
+                interimsFrameBuilder.addAllColumns(Arrays.asList(tempColumns));
+            }
+        }
+        Column[] replacementColumns = convertColumns(
+                inplaceOperationSimulator.getReplacementColumnNames(), 
+                inplaceOperationSimulator.getReplacementColumnTypes());
+        interimsFrameBuilder.addAllColumns(Arrays.asList(replacementColumns));
+        pipelineBuilder.setInterims(interimsFrameBuilder);
+        setOutputColumns(pipelineBuilder, lastStage, inplaceOperationSimulator);
+        Pipeline pipeline = pipelineBuilder.build();
+
+        return pipeline;
+    }
+    
+    public static MojoPipeline convert(Assembly assembly) throws IOException {
+        Pipeline pipeline = convertToProtoBufPipeline(assembly);
+        File tempFile = File.createTempFile("Pipeline", ".mojo");
+        tempFile.deleteOnExit();
+        ProtobufPipelineWriter writer = new ProtobufPipelineWriter(pipeline);
+        try (FileOutputStream outputStream = new FileOutputStream(tempFile)) {
+            writer.writeTo(outputStream);
+        }
+        ByteVec mojoData = NFSFileVec.make(tempFile);
+        return new MojoPipeline(mojoData);
+    }
+
+    private static Column convertColumn(String name, String type) {
+        Column.Builder builder = Column.newBuilder();
+        builder = builder.setName(name);
+        if (type.equals("Numeric")) {
+            builder.setFloat64Type(ColumnOuterClass.Float64Type.newBuilder().build());
+        } else {
+            builder.setStrType(ColumnOuterClass.StrType.newBuilder().build());
+        }
+        return builder.build();
+    }
+
+    private static Column[] convertColumns(String[] names, String[] types) {
+        if (names.length != types.length) {
+            throw new IllegalArgumentException(
+                String.format("The length of names and types must be the same, " +
+                        "but length of names is %d and length of types is %d.", names.length,  types.length));
+        }
+        Column[] result = new Column[names.length];
+        for (int i = 0; i < result.length; i++) {
+            result[i] = convertColumn(names[i], types[i]);
+        }
+        return result;
+    }
+
+    private static Transformation convertStage(Transform stage, InplaceOperationSimulator inplaceOperationSimulator){
+        if (stage instanceof H2OColSelect) {
+            return convertColSelect((H2OColSelect)stage, inplaceOperationSimulator);
+        } else if (stage instanceof H2OBinaryOp) {
+            return convertBinaryOp((H2OBinaryOp)stage, inplaceOperationSimulator);
+        } else if (stage instanceof H2OColOp) {
+            return convertColOp((H2OColOp)stage, inplaceOperationSimulator);
+        } else {
+            throw new UnsupportedOperationException(
```

Comment:
nitpik:  Could use a switch here if you want.

Reply:
Please can you give me an example? I thought that pattern matching (type comparison) was added to switch in java 17. https://openjdk.org/jeps/406